### PR TITLE
bpo-32734: Fix asyncio.Lock multiple acquire safety issue

### PR DESCRIPTION
**Pull request by bharel**
As promised, found and fixed in the same day due to the high importance.
This will cause almost non-existing overhead (one try statement) and eliminate the bug.

https://bugs.python.org/issue32734

### CODE REVIEW

File: Lib/test/test_asyncio/test_locks.py [hunk 1]
```diff
@@ -200,6 +200,42 @@ def test_cancel_race(self):
...
+  
+        async def lockandtrigger():
+            await lock.acquire()
+            self.loop.call_soon(trigger)
+          
+        def trigger():
+            t1.cancel()
+            lock.release()
+
+        asyncio.Task(lockandtrigger(), loop=self.loop)
```
Comment (asvetlov) [inline Lib/test/test_asyncio/test_locks.py]: Please use  `loop.create_task()` for a task creation

File: Lib/test/test_asyncio/test_locks.py [hunk 2]
```diff
@@ -200,6 +200,42 @@ def test_cancel_race(self):
...
+        # First loop acquires all
+        test_utils.run_briefly(self.loop)
+        # Second loop calls trigger
+        test_utils.run_briefly(self.loop)
+        # Third loop calls cancellation
+        test_utils.run_briefly(self.loop)
+
+        # Make sure only one lock was taken
+        self.assertEqual(lock_count, 1)
+
```
Comment (asvetlov) [inline Lib/test/test_asyncio/test_locks.py]: Please explicitly make sure that all four tasks are finished at the end of test

**Review by asvetlov — CHANGES_REQUESTED** [review]
Looks good, the change is correct.

Please fix minor issues for test.

File: Lib/test/test_asyncio/test_locks.py (hunk 1 above)
Comment (bharel) [reply Lib/test/test_asyncio/test_locks.py]: Done. Keep in mind asyncio.Task was used in the rest of the tests so I used it in order to be consistent.

File: Lib/test/test_asyncio/test_locks.py (hunk 2 above)
Comment (bharel) [reply Lib/test/test_asyncio/test_locks.py]: Done.

**Comment by bharel** [comment]
Alright. Done with changes and backports. Anything else I can help with? :-)

**Comment by bharel** [comment]
I have made the requested changes; please review again

File: Lib/asyncio/locks.py [hunk 3]
```diff
@@ -181,18 +181,22 @@ def locked(self):
             self._locked = True
             return True
 
+        _waiters = self._waiters
```
Comment (1st1) [inline Lib/asyncio/locks.py]: Nit: just use `self._waiters`. No need for a new variable.

File: Lib/asyncio/locks.py [hunk 4]
```diff
@@ -181,18 +181,22 @@ def locked(self):
...
         fut = self._loop.create_future()
-        self._waiters.append(fut)
+        _waiters.append(fut)
         try:
-            await fut
-            self._locked = True
-            return True
+            try:
```
Comment (1st1) [inline Lib/asyncio/locks.py:191]: This is very subtle.  Can you add a comment why we use two nested 'try' statements here? I.e. that it's important that 'finally' block is called before 'except CancelledError'

File: Lib/asyncio/locks.py [hunk 5]
```diff
@@ -212,11 +216,15 @@ def release(self):
             raise RuntimeError('Lock is not acquired.')
 
     def _wake_up_first(self):
-        """Wake up the first waiter who isn't cancelled."""
-        for fut in self._waiters:
-            if not fut.done():
-                fut.set_result(True)
-                break
+        """Wake up the first waiter if it isn't done."""
```
Comment (1st1) [inline Lib/asyncio/locks.py:221]: What if it is done?  Can there be a situation where you have [fut<done>, fut<not done>] in the _waiters list?

**Review by 1st1 — CHANGES_REQUESTED** [review]
I've left a few review comments

File: Lib/asyncio/locks.py (hunk 3 above)
Comment (bharel) [reply Lib/asyncio/locks.py]: New variable was created in order to avoid the double attribute lookup, I'll remove it if you think it's more readable though.

File: Lib/asyncio/locks.py (hunk 5 above)
Comment (bharel) [reply Lib/asyncio/locks.py:221]: There can be indeed, when 2 or more locks are waiting on acquire.
If the first future is done, it means that later on it will either get the lock, or it was cancelled in which case it will wake up the next waiter upon loop's next cycle.
The patch fixes the bug by avoiding multiple done waiters at once and cleaning up the waiters list asap.
The only way multiple done waiters can happen is if 2 futures get cancelled in which case only 1 waiter will wake up, and the second cancelled future will not wake up another one as lock would have already been taken (`if not self._locked: self._wake_up_first()`) or it wasn't taken yet it which case it's still sitting .done() in the waiter's list.

In summary:
.done necessarily means either the lock is immediately taken next loop or if it isn't, (.done occurred because of cancellation) next one wakes up. So if we have a done future not including the one currently running we shouldn't wake up another one as it will cause a double wakeup, hence, the almighty bug.

File: Lib/asyncio/locks.py [hunk 6]
```diff
@@ -181,18 +181,22 @@ def locked(self):
...
+        _waiters.append(fut)
         try:
-            await fut
-            self._locked = True
-            return True
+            try:
+                await fut
+            finally:
+                _waiters.remove(fut)
         except futures.CancelledError:
```
Comment (bharel) [inline Lib/asyncio/locks.py]: @1st1 @asvetlov I'll be honest, the only problem I have (which originates from the original code anyway) is what will happen if the exception isn't a CancelledError (such as `task.__step(self, exc=OutOfCoconutsException)`). Apart from Monty Python having no horse, the code will enter a deadlock as no one will wake up the next waiter. I'm not entirely sure what exceptions can pass into the task's __step as task doesn't allow set_exception but still better be safe than sorry no?

File: Lib/asyncio/locks.py (hunk 4 above)
Comment (bharel) [reply Lib/asyncio/locks.py:191]: Sure thing :-)

**Comment by bharel** [comment]
I have made the requested changes; please review again

**Review by asvetlov — APPROVED** [review]
LGTM.
@1st1 please review again

File: Lib/asyncio/locks.py (hunk 5 above)
Comment (1st1) [reply Lib/asyncio/locks.py:221]: OK, makes sense.  Could you please add this all as a comment to `_wake_up_first`? :)

Comment (bharel) [reply Lib/asyncio/locks.py:221]: Phew, now that's a long comment, lemme see how can I shorten it while still managing to fully explain it. Challenge accepted :P

Comment (1st1) [reply Lib/asyncio/locks.py:221]: Sure, if you have time to write a short comment :)

Sorry for being anal about this. Every once in a while we find a new bug in `asyncio.Lock` or `asyncio.Queue` and it's always a challenge to wrap your head around how they are actually working.  Every line of code is significant, sometimes it's not safe to do a simple refactoring because things break in very non-obvious ways.

**Review by 1st1 — APPROVED** [review]

File: Lib/asyncio/locks.py (hunk 5 above)
Comment (bharel) [reply Lib/asyncio/locks.py:221]: Nah, all good mate. It took some time to wrap my head around it as well.
Luckily I didn't actually encounter the bug at production but rather found it by reading the implementation which took a couple of hours ;-)

How's this comment? Attempting to be short and to the point as much as you can with synchronization.

**Comment by bharel** [comment]
Alright, I guess that's a green light for my backports?

**Comment by bharel** [comment]
Done with 3.6 backport.
Btw, doesn't it apply to 3.5 as well or did we stop supporting it?

**Comment by 1st1** [comment]
3.5 is in security-only fixes mode IIRC.
